response.go

```diff
@@ -0,0 +1,110 @@
+package typhon
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	log "github.com/mondough/slog"
+	"github.com/mondough/terrors"
+	"golang.org/x/net/context"
+)
+
+type Response struct {
+	*http.Response
+	Error error
+	ctx   context.Context
+}
+
+// Encode serialises the passed object as JSON into the body (and sets appropriate headers).
+func (r *Response) Encode(v interface{}) {
+	if err := json.NewEncoder(r).Encode(v); err != nil {
+		r.Error = terrors.Wrap(err, nil)
+		log.Warn(r.ctx, "Failed to encode response body: %v", err)
+		return
+	}
+	r.Header.Set("Content-Type", "application/json")
+}
+
+// Decode de-serialises the JSON body into the passed object.
+func (r *Response) Decode(v interface{}) error {
+	err := error(nil)
+	if r.Response == nil {
+		err = terrors.InternalService("", "Response has no body", nil)
+	} else {
+		err := json.NewDecoder(r.Body).Decode(v)
+		if err != nil {
+			log.Warn(r.ctx, "Failed to decode response body: %v", err)
+		}
+		err = terrors.WrapWithCode(err, nil, terrors.ErrBadResponse)
+	}
+	if r.Error == nil {
```

Comment:
In the case this is non-nil, but `r.Response` is nil we throw away the error without logging this.